File: 1283-find-the-smallest-divisor-given-a-threshold/1283-find-the-smallest-divisor-given-a-threshold.java
class Solution {
    public int smallestDivisor(int[] nums, int threshold) {
        return find(nums,threshold);
    }

    private static int result(int[] nums,int mid){
        int result=0;
        for(int num:nums){
            result += (num+mid-1)/mid; 
        }
        return result;
    }

    private static int find(int[] nums,int threshold){
        int left = 1;  
        int right = 0; 

        for(int num : nums){
            if(num > right){
                right = num;
            }
        }
        int ans=-1;

        while(left<=right){
            int mid = left+(right-left)/2;
            if(result(nums,mid)<=threshold){
                ans=mid;
                right=mid-1;
            }else{
                left=mid+1;
            }
        }
        return ans;
    }
}

File: 1283-find-the-smallest-divisor-given-a-threshold/1283-find-the-smallest-divisor-given-a-threshold.cpp
class Solution {
public:
    int smallestDivisor(vector<int>& nums, int threshold) {
        sort(nums.begin(),nums.end());
        int n=nums.size();
        int low=1;
        int high=nums[n-1];
        int ans=-1;
        while(low<=high){
            int mid=(high-low)/2+low;
            if(canDivide(nums,threshold,mid)){
                ans=mid;
                high=mid-1;
                
            }else{
                low=mid+1;
            }
        }
        return ans;
    }

    bool canDivide(vector<int>& nums,int threshold,int div){
        int ans=0;
        for(auto num:nums){
           ans+=ceil((double)(num)/(double)(div));
           if(ans>threshold){
            return false;
           }
        }
        return true;
    }
};